### ISSUE TITLE
[consistent-type-definitions] Don't automatically fix

### ISSUE THREAD

**Issue by Hoishin** [typescript-eslint/typescript-eslint#718]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/consistent-type-definitions": ["error"]
  }
}
```

TypeScript: https://www.typescriptlang.org/play/index.html?noUnusedLocals=true&noUnusedParameters=true&esModuleInterop=true#code/C4TwDgpgBAysBOBLAdgcwEoQMYHt4BMoBeKTXAgHgGcEVUAaKGpNAPgCh3RIoAxHHBQAqUCAA9gEZPiqxaaMnnytiUIZxST4AMwCGWaAEkAQrvhQA3uyg2ou+wC4oAcgBG75+wC+nbtCGm5iRWtnaOLu6unj5c4NAAcjjAAPIA1qr8goauZhx+UGkZAsKBrEA

**Expected Result**

It should **NOT** automatically convert between interface and type, either way, which causes type errors.

**Actual Result**

It automatically converts between interface and type.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.12.0` |
| `@typescript-eslint/parser`        | `1.12.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `10.16.0` |
| `npm`                              | `6.10.1` |

**Comment by bradzacher** [comment]
I'm not sure if this is something that should be fixed.
This rule is intending to enforce that you use consistent type definitions, which you are not, because you're using `type` so that you can use the side-effect of it having an implicit map.

The fix is simple - add `[k: string]: string` to your interface.

**Comment by Hoishin** [comment]
> enforce that you use consistent type definitions, which you are not
In the case where the first two definitions are in from a library and the rest are yours, even if you are completely consistent this will still be error.

> add `[k: string]: string`
That's much more different type and that doesn't sound the right solution. For instance, the `NotOk` type in the playground example would accept any string as key, not just `aaa`.
https://www.typescriptlang.org/play/index.html?noUnusedLocals=true&noUnusedParameters=true&esModuleInterop=true#code/C4TwDgpgBAysBOBLAdgcwEoQMYHt4BMoBeKTXAgHgGcEVUAaKGpNAPgCh3RIoAxHHBQAqUCAA9gEZPiqxaaMnnytiUIZxST4AMwCGWaAEkAQrvhQA3uyg2ou+wC4oAcgBG75wG5rtgNoBrJ2Y6AF0g+VR2AF9ObmghU3MSK1s7Rxd3V2do2PBoADkcYAB5f1V+QRMzDnxsABszaFxkGihkIpxAqEKS-3Zmqhw6iAA6OpxUAAp24E6RgCtFxYBKXJ5S8oFhRNYgA

**Comment by Hoishin** [comment]
The rule should just simply say you that you are inconsistent without the auto fix. The auto fix shouldn't cause type error like this.

**Comment by bradzacher** [comment]
Reading more into it, it's by design, and there's no way around it.
https://github.com/microsoft/TypeScript/issues/15300

If you absolutely 100% need to have a type definition there, then you can either:

convert it to the equivalent of a type by using `Pick`:
```ts
type InterfaceToType<T> = Pick<T, keyof T>;

type NotOk = Foo<InterfaceToType<IBar>>;
```

or disable the rule for that line so that other devs can see exactly why you can't use an interface
```ts
// interfaces cannot be assigned to Record types, so we have to use a type here
// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
type TBar = {
    aaa: 'bbb'
}
```

Like I said - this rule is about consistently defining your object types. If you need to use an implicit feature of `type`, but you have it configured to `interface`, then it is an edge case - which I'd suggest you use the disable comments for.

**Comment by Hoishin** [comment]
As I've said my point isn't the rule itself but the auto fix feature. If it's by design and no way around it, should it be converted automatically?

**Comment by bradzacher** [comment]
We straight up can't detect this case. We'd have to investigate your declarations across boundaries in order to discern whether or not each `interface` is used in a place that requires a `type`. This is a no-go perf wise.

**Comment by Hoishin** [comment]
Would it be too much for edge cases to remove auto fixer entirely for this rule?

**Comment by bradzacher** [comment]
Others may have opinions, but I'm of the opinion that you should be disabling the lint with a comment for the edge case to make it clear that you 100% intended to use a `type` instead of an `interface`.
In which case the fixer won't run at all.
